### PR TITLE
Allow control of finalization garbage collection with `.spec.deletionPolicy`

### PR DESCRIPTION
**Pull request by erikgb**
The `prune` setting of Kustomizations has a two-fold behavior. When set to `true` FLux will:
1. Ensure stale managed resources are deleted.
2. Garbage collect managed resources when Kustomization is deleted (using a finalizer).

We would like to have 1, but not 2, and this is currently not possible. This PR introduces a new API field allowing control over finalization garbage collection independently of `prune`.

### Some background/motivation for this proposed enhancement:

We are using Flux to provision review environments in a multi-tenant environment (shared cluster), and our tenants are not allowed to create Kubernetes namespaces directly. Instead they have to use a custom namespace-scoped resource as a controller for the namespace (and RBAC). This indirect management of namespaces makes it hard to create Kustomizations inside review namespaces. To overcome this, we use the `targetNamespace` field. This works well when the review namespace is created. But when the review namespace is teared down by our review environment automation, there is a finalization race condition between Flux and our namespace controller. If Flux looses this race, the Kustomization is stuck in non-finalized state (because the namespace is gone) and requires manual intervention from a cluster-admin. This is something we definitely would like to avoid! Since the namespace is controlled and eventually deleted by our namespace controller, no resources are in fact orphaned if we disable FLux garbage collection.

Orphaning resources is generally a bad idea, but is sometimes required. 😉 
~~Once we agree on the naming, I will submit a PR to update the docs in the [website repo](https://github.com/fluxcd/website).~~

Close https://github.com/fluxcd/kustomize-controller/issues/662

### CODE REVIEW

File: api/v1/kustomization_types.go
Comment (stefanprodan) [inline api/v1/kustomization_types.go:104]:
```suggestion
	// Valid values are ('MirrorPrune' (default), 'Delete', 'Orphan')
```

**Review by stefanprodan — COMMENTED** [review]
Please add a new section to the API docs right after https://github.com/fluxcd/kustomize-controller/blob/main/docs/spec/v1/kustomizations.md#prune

File: internal/controller/kustomization_controller.go
Comment (stefanprodan) [inline internal/controller/kustomization_controller.go:959]:
```suggestion
func finalizerShouldDeleteResources(obj *kustomizev1.Kustomization) bool {
```

File: internal/controller/kustomization_controller.go
Comment (stefanprodan) [inline internal/controller/kustomization_controller.go:962]: Please add a function to the API package called `GetDeletionPolicy()`, similar to `GetRetryInterval()`, that returns the default value when the field value is set to string empty. Then change the logic here to `if obj.GetDeletionPolicy() == kustomizev1.DeletionPolicyMirrorPrune`.

File: api/v1/kustomization_types.go
Comment (stefanprodan) [inline api/v1/kustomization_types.go:107]:
```suggestion
	DeletionPolicy string `json:"deletionPolicy,omitempty"`
```

No need to set a pointer to a string.

File: internal/controller/kustomization_deletion_policy_test.go
Comment (matheuscscp) [inline internal/controller/kustomization_deletion_policy_test.go:39]: Please test also `MirrorPrune` with the different possible values for `prune`.

**Review by matheuscscp — COMMENTED** [review]
Thanks for this contribution! I think it's a really valuable one.

File: internal/controller/kustomization_deletion_policy_test.go
Comment (erikgb) [reply internal/controller/kustomization_deletion_policy_test.go:39]: @matheuscscp, will do! Are there any recommendations or established patterns in this project to avoid so much duplicated test code? If not, I am pretty good at copy-modify. 😉

File: internal/controller/kustomization_controller.go
Comment (erikgb) [reply internal/controller/kustomization_controller.go]: This should be fixed now. I will write docs and add more tests before rebasing and squashing everything into a single commit.

File: internal/controller/kustomization_deletion_policy_test.go
Comment (matheuscscp) [reply internal/controller/kustomization_deletion_policy_test.go:38]: We always try to do the usual table testing widespread in Go projects. However if the test cases are too different then it's OK to have separate `func TestXXX` tests for them. In this specific case you could for example group together the cases where deletion is expected, and make another group where orphaning is expected.

Comment (erikgb) [reply internal/controller/kustomization_deletion_policy_test.go:38]: Thanks! Do you have a good example of how you write table-driven tests in Flux projects? Assuming there are multiple ways to do this.

File: docs/spec/v1/kustomizations.md
Comment (tenstad) [inline docs/spec/v1/kustomizations.md:194]: Time of completion is what's relevant?

```suggestion
before Flux completes the Kustomization finalization.
```

File: api/v1/kustomization_types.go
Comment (tenstad) [reply api/v1/kustomization_types.go:104]: The mapping could be described in the docs instead, that might be helpful (and sufficient).

File: internal/controller/kustomization_deletion_policy_test.go
Comment (erikgb) [reply internal/controller/kustomization_deletion_policy_test.go:38]: Thanks! I have updated the tests now. I have some issues running the tests on my workstation (unrelated to this change). I get the following error:

````
panic: Failed to create Vault instance: could not connect to docker: reached retry deadline: Get "http://127.0.0.1:32769/v1/sys/init": dial tcp 127.0.0.1:32769: connect: connection refused
````

So would it be possible to approve workflows to check the tests (and CI in general)?

I also want to rebase this PR before merge. But I'll wait until it's "approved" (almost ready to merge).

File: api/v1/kustomization_types.go
Comment (erikgb) [reply api/v1/kustomization_types.go:104]: > I do like the convention of a final Defaults to statement

I agree in general. But the default is a bit "special" in this case.

File: docs/spec/v1/kustomizations.md
Comment (stefanprodan) [inline docs/spec/v1/kustomizations.md:182]: We need to add MirrorPrune first and explain what it does.

File: docs/spec/v1/kustomizations.md
Comment (stefanprodan) [inline docs/spec/v1/kustomizations.md:196]:
```suggestion
For special cases when the managed resources are removed by other means (e.g. the deletion of the namespace specified with `.spec.targetNamespace`), you can set the deletion policy to `Orphan`:
```

File: api/v1/kustomization_types.go
Comment (tenstad) [reply api/v1/kustomization_types.go:104]: Don't see the harm in writing the default _value_ (not behaviour) when the sentence before it describes the behavior of that value.

I think the connection between the two current statements can be improved. _mirror value of Prune_ relates to _MirrorPrune_ and _unspecified_ relates to _default_, but it is not that explicit, which can lead to confusion.

> Will mirror value of Prune if unspecified.

> Valid values are ('MirrorPrune' (default), 'Delete', 'Orphan')

File: docs/spec/v1/kustomizations.md
Comment (erikgb) [reply docs/spec/v1/kustomizations.md:184]: Made an attempt. Please take a look.

File: internal/controller/kustomization_deletion_policy_test.go
Comment (stefanprodan) [inline internal/controller/kustomization_deletion_policy_test.go:2]:
```suggestion
Copyright 2024 The Flux authors
```

**Comment by erikgb** [comment]
Ok, it seems like we are approaching something here. 😸  I have just rebased and squashed all commits. The PR should be ready to merge if you are satisfied with the proposed changes.

File: docs/spec/v1/kustomizations.md
Comment (tenstad) [inline docs/spec/v1/kustomizations.md:174]:
```suggestion
`.spec.deletionPolicy` is an optional field that allows control over
```

File: docs/spec/v1/kustomizations.md
Comment (tenstad) [inline docs/spec/v1/kustomizations.md:176]: Not totally sure about this one, but it's not the concrete true/false value that is mirrored. Should that be addressed here?

```suggestion
is to mirror the configuration of [`.spec.prune`](#prune).
```

File: docs/spec/v1/kustomizations.md
Comment (tenstad) [inline docs/spec/v1/kustomizations.md:184]: Keep it simple and non-technical, and equal between the two sentances?
```suggestion
   is deleted.
- `Orphan` - Leave the managed resources when the Kustomization is deleted.
```

File: docs/spec/v1/kustomizations.md
Comment (tenstad) [inline docs/spec/v1/kustomizations.md:197]: Not relevant here, might as well remove?
```suggestion
```

File: api/v1/kustomization_types.go
Comment (tenstad) [inline api/v1/kustomization_types.go:105]: If becoming a bit too technical and type-specific, writing it out in text can also work. Up to you!

```suggestion
	// 'Orphan'). 'MirrorPrune' mirrors the Prune field (orphan if false,
	// delete if true). Defaults to 'MirrorPrune'.
```

**Review by tenstad — COMMENTED** [review]
Looking good! Just some minor nitpicks/thouhts regarding text.

File: docs/spec/v1/kustomizations.md
Comment (matheuscscp) [reply docs/spec/v1/kustomizations.md:197]: I'm ok with removing it but I'd prefer to keep it, otherwise a new user reading this quickly might think `Kustomization` is cluster-scoped. :shrug:

File: docs/spec/v1/kustomizations.md
Comment (matheuscscp) [inline docs/spec/v1/kustomizations.md:199]:
```suggestion
```

File: internal/controller/kustomization_deletion_policy_test.go
Comment (matheuscscp) [inline internal/controller/kustomization_deletion_policy_test.go:130]: I believe we can move all these assertions outside the test case. The test cases should be only the ones in the table on the beginning.

File: internal/controller/kustomization_deletion_policy_test.go
Comment (matheuscscp) [inline internal/controller/kustomization_deletion_policy_test.go:134]: Ditto.

File: internal/controller/kustomization_deletion_policy_test.go
Comment (matheuscscp) [inline internal/controller/kustomization_deletion_policy_test.go:240]: Ditto.

File: internal/controller/kustomization_deletion_policy_test.go
Comment (matheuscscp) [inline internal/controller/kustomization_deletion_policy_test.go:244]: Ditto.

File: internal/controller/kustomization_deletion_policy_test.go
Comment (matheuscscp) [inline internal/controller/kustomization_deletion_policy_test.go:148]: Looking at these two top-level tests I think the difference between them is not huge. I think we could move all the cases into a single top-level test function and add a boolean to the test case struct to indicate whether deletion is expected or not and have and if-else that tests both cases at the end of the test code. Sorry for misleading you on the opposite direction before!

**Review by matheuscscp — COMMENTED** [review]
Great stuff, I'm ready to give my approval after all the open discussions are resolved!

File: internal/controller/kustomization_deletion_policy_test.go
Comment (erikgb) [reply internal/controller/kustomization_deletion_policy_test.go:148]: No, problem. I will fix it! The same thought struck me when changing this.

File: docs/spec/v1/kustomizations.md
Comment (erikgb) [reply docs/spec/v1/kustomizations.md:199]: This was copied from another example in the docs. Without it, a reader might wonder if this is a valid Kustomization.

File: internal/controller/kustomization_deletion_policy_test.go
Comment (erikgb) [reply internal/controller/kustomization_deletion_policy_test.go]: Done

**Comment by erikgb** [comment]
@matheuscscp All discussions should now be resolved. Can you please approve CI to see if I made any mistakes in the test refactoring?

**Review by matheuscscp — APPROVED** [review]
Great contribution! Thanks @erikgb

**Review by tenstad — APPROVED** [review]

**Review by stefanprodan — APPROVED** [review]
LGTM

Thanks @erikgb 🏅

**Pull request merged**
